File: src/xpict/draw/svg.py
"""Serialize Scene → SVG / HTML.

Text primitives are drawn as glyph **paths** (Liberation Sans outlines), not
``<text>``/``<tspan>``. Markup (``\\alpha``, ``**bold**``, …) is expanded when
outlining so bold/italic pick the matching bundled face.
"""

from __future__ import annotations

import html
from xml.etree.ElementTree import Element, SubElement, tostring

from xpict.contracts.scene import (
    CirclePrim,
    PathPrim,
    Primitive,
    Scene,
    TextPrim,
    Viewport,
)
from xpict.draw.glyphs import compile_text_path_d
from xpict.draw.richtext import StyledText


def _render_primitive(parent: Element, prim: Primitive) -> None:
    if isinstance(prim, PathPrim):
        attrs = {
            "d": prim.d,
            "fill": prim.fill or "none",
            "stroke": prim.stroke or "none",
            "stroke-width": str(prim.stroke_width),
            "stroke-linecap": prim.stroke_linecap or "round",
            "stroke-linejoin": "round",
            "opacity": str(prim.opacity),
        }
        if prim.stroke_dasharray:
            attrs["stroke-dasharray"] = prim.stroke_dasharray
        if prim.cls:
            attrs["class"] = prim.cls
        SubElement(parent, "path", attrs)
    elif isinstance(prim, CirclePrim):
        attrs = {
            "cx": f"{prim.cx:.2f}",
            "cy": f"{prim.cy:.2f}",
            "r": f"{prim.r:.2f}",
            "opacity": str(prim.opacity),
            "fill": prim.fill or "none",
        }
        if prim.stroke:
            attrs["stroke"] = prim.stroke
            attrs["stroke-width"] = str(prim.stroke_width)
        if prim.cls:
            attrs["class"] = prim.cls
        SubElement(parent, "circle", attrs)
    elif isinstance(prim, TextPrim):
        # Markup → StyledText (Unicode + spans) → glyph path shapes.
        styled = StyledText.from_markup(prim.text)
        d = compile_text_path_d(
            styled,
            prim.x,
            prim.y,
            font_size=prim.font_size,
            anchor=prim.anchor,
        )
        if not d:
            return
        attrs = {
            "d": d,
            "fill": prim.fill,
            "stroke": "none",
            "opacity": "1",
            "data-text": styled.text,
        }
        if prim.cls:
            attrs["class"] = prim.cls
        SubElement(parent, "path", attrs)


def _render_viewport(
    parent: Element,
    vp: Viewport,
    *,
    layers: tuple[str, ...] | None = None,
) -> None:
    g = SubElement(
        parent, "g", {"class": "xpict-mol", "transform": f"translate({vp.x},{vp.y})"}
    )
    if vp.id:
        g.set("data-id", vp.id)
    want = set(layers) if layers is not None else None
    for layer in vp.layers:
        if want is not None and layer.name not in want:
            continue
        if not layer.primitives:
            continue
        lg = SubElement(g, "g", {"class": f"layer-{layer.name}", "id": layer.name})
        for prim in layer.primitives:
            _render_primitive(lg, prim)


def _fmt_user(value: float) -> str:
    """Format a user-unit length (viewBox / geometry), trimming float noise."""
    return f"{value:.4f}".rstrip("0").rstrip(".") or "0"


def _fmt_css_px(value: float) -> str:
    """Format an intrinsic CSS size so browsers lay the SVG out at content size."""
    return f"{_fmt_user(value)}px"


def scene_to_svg(scene: Scene) -> str:
    # Explicit px width/height give a stable intrinsic size. viewBox stays in
    # user units so content scales only when CSS intentionally overrides size.
    w = _fmt_user(scene.width)
    h = _fmt_user(scene.height)
    root = Element(
        "svg",
        {
            "xmlns": "http://www.w3.org/2000/svg",
            "width": _fmt_css_px(scene.width),
            "height": _fmt_css_px(scene.height),
            "viewBox": f"0 0 {w} {h}",
            "class": "xpict",
        },
    )
    # Shade under the document halo; ink above it.
    for vp in scene.viewports:
        _render_viewport(root, vp, layers=("shading",))
    if scene.halo:
        hg = SubElement(root, "g", {"class": "xpict-halo", "id": "halo"})
        for prim in scene.halo:
            _render_primitive(hg, prim)
    for vp in scene.viewports:
        _render_viewport(root, vp, layers=("bonds", "labels", "marks", "overlay"))
    if scene.overlays:
        og = SubElement(root, "g", {"class": "xpict-overlays", "id": "overlays"})
        for prim in scene.overlays:
            _render_primitive(og, prim)
    return '<?xml version="1.0" encoding="UTF-8"?>\n' + tostring(root, encoding="unicode")


def scene_to_html(scene: Scene, *, title: str | None = None) -> str:
    svg = scene_to_svg(scene)
    if svg.startswith("<?xml"):
        svg = svg.split("\n", 1)[1]
    t = html.escape(title or "xpict")
    return f"""<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>{t}</title>
<style>
  body {{ margin: 0; font-family: system-ui, sans-serif; background: #fafafa; }}
  .xpict-page {{ padding: 1rem; box-sizing: border-box; }}
  /* Honor SVG width/height attrs (intrinsic size). Shrink only if the page is
     narrower — never stretch separate depictions up to a shared column width. */
  .xpict-page svg.xpict {{
    display: inline-block;
    vertical-align: middle;
    width: auto;
    height: auto;
    max-width: 100%;
    flex: 0 0 auto;
  }}
  .xpict-row {{
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    gap: 1rem;
  }}
</style>
</head>
<body>
<main class="xpict-page">
{svg}
</main>
</body>
</html>
"""

File: crates/xpict/src/svg.rs
//! Scene → SVG string (parity with JS ``sceneToSvg`` / Python ``scene_to_svg``).

use xpict_core::scene::{LayerName, Primitive, Scene, TextAnchor, Viewport};

fn esc(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('"', "&quot;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
}

fn fmt(n: f64) -> String {
    if !n.is_finite() {
        return "0".into();
    }
    let t = (n * 100.0).round() / 100.0;
    let s = format!("{t}");
    if s == "-0" {
        "0".into()
    } else {
        s
    }
}

fn fmt_path_d(d: &str) -> String {
    let mut out = String::with_capacity(d.len());
    let bytes = d.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        let c = bytes[i] as char;
        let start_num = c == '-' || c == '+' || c == '.' || c.is_ascii_digit();
        if start_num {
            let start = i;
            i += 1;
            while i < bytes.len() {
                let ch = bytes[i] as char;
                if ch.is_ascii_digit() || ch == '.' {
                    i += 1;
                    continue;
                }
                if (ch == 'e' || ch == 'E') && i + 1 < bytes.len() {
                    let n = bytes[i + 1] as char;
                    if n == '+' || n == '-' || n.is_ascii_digit() {
                        i += 2;
                        while i < bytes.len() && (bytes[i] as char).is_ascii_digit() {
                            i += 1;
                        }
                        break;
                    }
                }
                break;
            }
            let token = &d[start..i];
            if let Ok(v) = token.parse::<f64>() {
                out.push_str(&fmt(v));
            } else {
                out.push_str(token);
            }
        } else {
            out.push(c);
            i += 1;
        }
    }
    out
}

fn attr(name: &str, value: Option<&str>) -> String {
    match value {
        Some(v) if !v.is_empty() => format!(" {name}=\"{}\"", esc(v)),
        _ => String::new(),
    }
}

fn attr_num(name: &str, value: f64) -> String {
    format!(" {name}=\"{}\"", fmt(value))
}

fn render_primitive(p: &Primitive) -> String {
    match p {
        Primitive::Path {
            d,
            stroke,
            fill,
            stroke_width,
            opacity,
            stroke_dasharray,
            stroke_linecap,
            class,
            data_text,
        } => format!(
            "<path d=\"{}\"{}{}{}{}{}{}{}{}{}/>",
            fmt_path_d(d),
            attr("fill", Some(fill.as_deref().unwrap_or("none"))),
            attr("stroke", stroke.as_deref()),
            attr_num("stroke-width", *stroke_width),
            attr(
                "stroke-linecap",
                Some(stroke_linecap.as_deref().unwrap_or("round"))
            ),
            " stroke-linejoin=\"round\"",
            attr_num("opacity", *opacity),
            attr("stroke-dasharray", stroke_dasharray.as_deref()),
            attr("class", class.as_deref()),
            attr("data-text", data_text.as_deref()),
        ),
        Primitive::Circle {
            cx,
            cy,
            r,
            fill,
            stroke,
            stroke_width,
            opacity,
            class,
        } => format!(
            "<circle cx=\"{}\" cy=\"{}\" r=\"{}\"{}{}{}{}{}/>",
            fmt(*cx),
            fmt(*cy),
            fmt(*r),
            attr("fill", Some(fill.as_deref().unwrap_or("none"))),
            attr("stroke", stroke.as_deref()),
            attr_num("stroke-width", *stroke_width),
            attr_num("opacity", *opacity),
            attr("class", class.as_deref()),
        ),
        Primitive::Text {
            x,
            y,
            text,
            fill,
            font_size,
            anchor,
            class,
        } => {
            let anchor_s = match anchor {
                TextAnchor::Start => "start",
                TextAnchor::Middle => "middle",
                TextAnchor::End => "end",
            };
            format!(
                "<text x=\"{}\" y=\"{}\" fill=\"{}\" font-size=\"{}\" \
                 font-family=\"Liberation Sans, Arial, sans-serif\" text-anchor=\"{anchor_s}\" \
                 dominant-baseline=\"alphabetic\"{}>{}</text>",
                fmt(*x),
                fmt(*y),
                esc(fill),
                fmt(*font_size),
                attr("class", class.as_deref()),
                esc(text),
            )
        }
    }
}

fn layer_name(n: LayerName) -> &'static str {
    match n {
        LayerName::Shading => "shading",
        LayerName::Halo => "halo",
        LayerName::Bonds => "bonds",
        LayerName::Labels => "labels",
        LayerName::Marks => "marks",
        LayerName::Overlay => "overlay",
    }
}

fn render_viewport_layers(vp: &Viewport, names: &[&str]) -> String {
    let mut chunks = String::new();
    for layer in &vp.layers {
        let name = layer_name(layer.name);
        if !names.contains(&name) || layer.primitives.is_empty() {
            continue;
        }
        let body: String = layer.primitives.iter().map(render_primitive).collect();
        chunks.push_str(&format!(
            "<g class=\"xpict-layer xpict-{name}\">{body}</g>"
        ));
    }
    chunks
}

/// Serialize a [`Scene`] to SVG (width/height = viewBox = SCALE units).
pub fn scene_to_svg(scene: &Scene) -> String {
    let w = fmt(scene.width);
    let h = fmt(scene.height);
    let mut parts = String::new();
    for vp in &scene.viewports {
        parts.push_str(&render_viewport_layers(vp, &["shading"]));
    }
    if !scene.halo.is_empty() {
        let body: String = scene.halo.iter().map(render_primitive).collect();
        parts.push_str(&format!("<g class=\"xpict-halo\" id=\"halo\">{body}</g>"));
    }
    for vp in &scene.viewports {
        parts.push_str(&render_viewport_layers(
            vp,
            &["bonds", "labels", "marks", "overlay"],
        ));
    }
    if !scene.overlays.is_empty() {
        let body: String = scene.overlays.iter().map(render_primitive).collect();
        parts.push_str(&format!(
            "<g class=\"xpict-overlays\" id=\"overlays\">{body}</g>"
        ));
    }
    format!(
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n\
         <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"{w}\" height=\"{h}\" \
         viewBox=\"0 0 {w} {h}\" class=\"xpict\" style=\"background:transparent\">{parts}</svg>"
    )
}
